// @comunica/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FederatedQuadSource = void 0;
const bus_rdf_resolve_quad_pattern_1 = require("@comunica/bus-rdf-resolve-quad-pattern");
const data_factory_1 = require("@comunica/data-factory");
const asynciterator_1 = require("asynciterator");
const rdf_data_factory_1 = require("rdf-data-factory");
const rdf_terms_1 = require("rdf-terms");
const sparqlalgebrajs_1 = require("sparqlalgebrajs");
const DF = new rdf_data_factory_1.DataFactory();
/**
 * A FederatedQuadSource can evaluate quad pattern queries over the union of different heterogeneous sources.
 * It will call the given mediator to evaluate each quad pattern query separately.
 */
class FederatedQuadSource {
    constructor(mediatorResolveQuadPattern, context, emptyPatterns, skipEmptyPatterns) {
        this.mediatorResolveQuadPattern = mediatorResolveQuadPattern;
        this.sources = context.get(bus_rdf_resolve_quad_pattern_1.KEY_CONTEXT_SOURCES);
        this.contextDefault = context.delete(bus_rdf_resolve_quad_pattern_1.KEY_CONTEXT_SOURCES);
        this.emptyPatterns = emptyPatterns;
        this.sourceIds = new Map();
        this.skipEmptyPatterns = skipEmptyPatterns;
        this.algebraFactory = new sparqlalgebrajs_1.Factory();
        // Initialize sources in the emptyPatterns datastructure
        if (this.skipEmptyPatterns) {
            for (const source of this.sources) {
                if (!this.emptyPatterns.has(source)) {
                    this.emptyPatterns.set(source, []);
                }
            }
        }
    }
    /**
     * Check if the given RDF term is not bound to an exact value.
     * I.e., if it is not a Variable.
     * @param {RDF.Term} term An RDF term.
     * @return {boolean} If it is not bound.
     */
    static isTermBound(term) {
        return term.termType !== 'Variable';
    }
    /**
     * Checks if the given (child) pattern is a more bound version of the given (parent) pattern.
     * This will also return true if the patterns are equal.
     * @param {RDF.BaseQuad} child A child pattern.
     * @param {RDF.BaseQuad} parent A parent pattern.
     * @return {boolean} If child is a sub-pattern of parent
     */
    static isSubPatternOf(child, parent) {
        return (!FederatedQuadSource.isTermBound(parent.subject) || parent.subject.equals(child.subject)) &&
            (!FederatedQuadSource.isTermBound(parent.predicate) || parent.predicate.equals(child.predicate)) &&
            (!FederatedQuadSource.isTermBound(parent.object) || parent.object.equals(child.object)) &&
            (!FederatedQuadSource.isTermBound(parent.graph) || parent.graph.equals(child.graph));
    }
    /**
     * If the given term is a blank node, return a deterministic named node for it
     * based on the source id and the blank node value.
     * @param term Any RDF term.
     * @param sourceId A source identifier.
     * @return If the given term was a blank node, this will return a skolemized named node, otherwise the original term.
     */
    static skolemizeTerm(term, sourceId) {
        if (term.termType === 'BlankNode') {
            return new data_factory_1.BlankNodeScoped(`bc_${sourceId}_${term.value}`, DF.namedNode(`${FederatedQuadSource.SKOLEM_PREFIX}${sourceId}:${term.value}`));
        }
        return term;
    }
    /**
     * Skolemize all terms in the given quad.
     * @param quad An RDF quad.
     * @param sourceId A source identifier.
     * @return The skolemized quad.
     */
    static skolemizeQuad(quad, sourceId) {
        return rdf_terms_1.mapTerms(quad, term => FederatedQuadSource.skolemizeTerm(term, sourceId));
    }
    /**
     * If a given term was a skolemized named node for the given source id,
     * deskolemize it again to a blank node.
     * If the given term was a skolemized named node for another source, return false.
     * If the given term was not a skolemized named node, return the original term.
     * @param term Any RDF term.
     * @param sourceId A source identifier.
     */
    static deskolemizeTerm(term, sourceId) {
        if (term.termType === 'BlankNode' && 'skolemized' in term) {
            term = term.skolemized;
        }
        if (term.termType === 'NamedNode') {
            if (term.value.startsWith(FederatedQuadSource.SKOLEM_PREFIX)) {
                const colonSeparator = term.value.indexOf(':', FederatedQuadSource.SKOLEM_PREFIX.length);
                const termSourceId = term.value.slice(FederatedQuadSource.SKOLEM_PREFIX.length, colonSeparator);
                // We had a skolemized term
                if (termSourceId === sourceId) {
                    // It came from the correct source
                    const termLabel = term.value.slice(colonSeparator + 1, term.value.length);
                    return DF.blankNode(termLabel);
                }
                // It came from a different source
                return null;
            }
        }
        return term;
    }
    /**
     * If the given source is guaranteed to produce an empty result for the given pattern.
     *
     * This prediction is done based on the 'emptyPatterns' datastructure that is stored within this actor.
     * Every time an empty pattern is passed, this pattern is stored in this datastructure for this source.
     * If this pattern (or a more bound pattern) is queried, we know for certain that it will be empty again.
     * This is under the assumption that sources will remain static during query evaluation.
     *
     * @param {IQuerySource} source
     * @param {RDF.BaseQuad} pattern
     * @return {boolean}
     */
    isSourceEmpty(source, pattern) {
        if (!this.skipEmptyPatterns) {
            return false;
        }
        const emptyPatterns = this.emptyPatterns.get(source);
        if (emptyPatterns) {
            for (const emptyPattern of emptyPatterns) {
                if (FederatedQuadSource.isSubPatternOf(pattern, emptyPattern)) {
                    return true;
                }
            }
        }
        return false;
    }
    /**
     * Get the unique, deterministic id for the given source.
     * @param source A data source.
     * @return The id of the given source.
     */
    getSourceId(source) {
        let sourceId = this.sourceIds.get(source);
        if (sourceId === undefined) {
            sourceId = `${this.sourceIds.size}`;
            this.sourceIds.set(source, sourceId);
        }
        return sourceId;
    }
    match(subject, predicate, object, graph) {
        // Counters for our metadata
        const metadata = { totalItems: 0 };
        let remainingSources = this.sources.length;
        // Anonymous function to handle totalItems from metadata
        const checkEmitMetadata = (currentTotalItems, source, pattern, lastMetadata) => {
            if (this.skipEmptyPatterns && !currentTotalItems) {
                // Because another call may have added more information in the meantime
                if (pattern && !this.isSourceEmpty(source, pattern)) {
                    this.emptyPatterns.get(source).push(pattern);
                }
            }
            if (!remainingSources) {
                if (lastMetadata && this.sources.length === 1) {
                    // If we only had one source, emit the metadata as-is.
                    it.setProperty('metadata', lastMetadata);
                }
                else {
                    it.setProperty('metadata', metadata);
                }
            }
        };
        const proxyIt = Promise.all(this.sources.map(async (source) => {
            const sourceId = this.getSourceId(source);
            // Deskolemize terms, so we send the original blank nodes to each source.
            // Note that some sources may not match bnodes by label. SPARQL endpoints for example consider them variables.
            const patternS = FederatedQuadSource.deskolemizeTerm(subject, sourceId);
            const patternP = FederatedQuadSource.deskolemizeTerm(predicate, sourceId);
            const patternO = FederatedQuadSource.deskolemizeTerm(object, sourceId);
            const patternG = FederatedQuadSource.deskolemizeTerm(graph, sourceId);
            let pattern;
            // Prepare the context for this specific source
            let context = bus_rdf_resolve_quad_pattern_1.getDataSourceContext(source, this.contextDefault);
            context = context.set(bus_rdf_resolve_quad_pattern_1.KEY_CONTEXT_SOURCE, { type: bus_rdf_resolve_quad_pattern_1.getDataSourceType(source), value: bus_rdf_resolve_quad_pattern_1.getDataSourceValue(source) });
            let output;
            // If any of the deskolemized blank nodes originate from another source,
            // or if we can predict that the given source will have no bindings for the given pattern,
            // return an empty iterator.
            if (!patternS || !patternP || !patternO || !patternG ||
                // eslint-disable-next-line no-cond-assign
                this.isSourceEmpty(source, pattern = this.algebraFactory
                    .createPattern(patternS, patternP, patternO, patternG))) {
                output = { data: new asynciterator_1.ArrayIterator([], { autoStart: false }) };
                output.data.setProperty('metadata', { totalItems: 0 });
            }
            else {
                output = await this.mediatorResolveQuadPattern.mediate({ pattern, context });
            }
            // Handle the metadata from this source
            output.data.getProperty('metadata', (subMetadata) => {
                if ((!subMetadata.totalItems && subMetadata.totalItems !== 0) || !Number.isFinite(subMetadata.totalItems)) {
                    // We're already at infinite, so ignore any later metadata
                    metadata.totalItems = Infinity;
                    remainingSources = 0;
                    checkEmitMetadata(Infinity, source, pattern, subMetadata);
                }
                else {
                    metadata.totalItems += subMetadata.totalItems;
                    remainingSources--;
                    checkEmitMetadata(subMetadata.totalItems, source, pattern, subMetadata);
                }
            });
            // Determine the data stream from this source
            let data = output.data.map(quad => FederatedQuadSource.skolemizeQuad(quad, sourceId));
            // SPARQL query semantics allow graph variables to only match with named graphs, excluding the default graph
            if (graph.termType === 'Variable') {
                data = data.filter(quad => quad.graph.termType !== 'DefaultGraph');
            }
            return data;
        }));
        // Take the union of all source streams
        const it = new asynciterator_1.TransformIterator(async () => new asynciterator_1.UnionIterator(await proxyIt), { autoStart: false });
        // If we have 0 sources, immediately emit metadata
        if (this.sources.length === 0) {
            it.setProperty('metadata', metadata);
        }
        return it;
    }
}
exports.FederatedQuadSource = FederatedQuadSource;
FederatedQuadSource.SKOLEM_PREFIX = 'urn:comunica_skolem:source_';
