// sparqlee/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.orderTypes = exports.isExpressionError = exports.ExpressionError = exports.AggregateEvaluator = exports.SyncEvaluator = exports.AsyncEvaluator = void 0;
var AsyncEvaluator_1 = require("./lib/evaluators/AsyncEvaluator");
Object.defineProperty(exports, "AsyncEvaluator", { enumerable: true, get: function () { return AsyncEvaluator_1.AsyncEvaluator; } });
var SyncEvaluator_1 = require("./lib/evaluators/SyncEvaluator");
Object.defineProperty(exports, "SyncEvaluator", { enumerable: true, get: function () { return SyncEvaluator_1.SyncEvaluator; } });
var AggregateEvaluator_1 = require("./lib/evaluators/AggregateEvaluator");
Object.defineProperty(exports, "AggregateEvaluator", { enumerable: true, get: function () { return AggregateEvaluator_1.AggregateEvaluator; } });
var Errors_1 = require("./lib/util/Errors");
Object.defineProperty(exports, "ExpressionError", { enumerable: true, get: function () { return Errors_1.ExpressionError; } });
Object.defineProperty(exports, "isExpressionError", { enumerable: true, get: function () { return Errors_1.isExpressionError; } });
var Ordering_1 = require("./lib/util/Ordering");
Object.defineProperty(exports, "orderTypes", { enumerable: true, get: function () { return Ordering_1.orderTypes; } });
